src/address-manager/index.js

```diff
@@ -0,0 +1,73 @@
+'use strict'
+
+const debug = require('debug')
+const log = debug('libp2p:addresses')
+log.error = debug('libp2p:addresses:error')
+
+const MultiaddrSet = require('./multiaddr-set')
+
+/**
+ * Responsible for managing known peers, as well as their addresses, protocols and metadata.
+ */
+class AddressManager {
+  /**
+   * @constructor
+   * @param {object} [options]
+   * @param {Array<string>} [options.listen = []] list of multiaddrs string representation to listen.
+   * @param {Array<string>} [options.announce = []] list of multiaddrs string representation to announce.
+   * @param {Array<string>} [options.noAnnounce = []] list of multiaddrs string representation to not announce.
+   */
+  constructor ({ listen = [], announce = [], noAnnounce = [] } = {}) {
+    this._listen = new MultiaddrSet(listen)
+    this._announce = new MultiaddrSet(announce)
+    this._noAnnounce = new MultiaddrSet(noAnnounce)
+  }
+
+  /**
+   * Get peer addresses to listen.
+   * @return {Array<Multiaddr>}
+   */
+  get listen () {
+    return this._listen.toArray()
+  }
+
+  /**
+   * Get peer addresses to announce.
+   * @return {Array<Multiaddr>}
+   */
+  get announce () {
+    return this._listen.toArray()
+      .concat(this._announce.toArray()) // Concat announce amendment
+      .filter((ma) => !this._noAnnounce.has(ma)) // Remove no announce addresses
+  }
+
+  /**
+   * Get peer addresses to not announce.
+   * @return {Array<Multiaddr>}
+   */
+  get noAnnounce () {
+    return this._noAnnounce.toArray()
+  }
+
+  /**
+   * Replace listen multiaddrs after trying to listen the original ones by the transports.
+   * This should also forward the replace side effects to the NoAnnounce sets.
+   * @param {Array<Multiaddr>} newMultiaddrs
+   * @return {Array<Multiaddr>}
+   */
+  _replaceListen (newMultiaddrs) {
```

Comment:
I don't think we should do this, it makes tracking things more difficult and since we're not tracking everything in `PeerInfo` this is a bit easier to manage. We should keep the `listen`, `announce`, and `noAnnounce` untouched. They could be added to, but we shouldn't internally replace them, these should be user controlled.

We should have a method that computes everything for advertised addresses, this could live on libp2p itself. `libp2p.getAddrs()` or similar. This method would do something like:
- Get the TransportManager addresses, `libp2p.transportManager.getAddrs()`
- Add in the announce addresses, `[...libp2p.transportManager.getAddrs(), ...libp2p.addressManager.announce]`
- And then filter out matching noAnnounce and any duplicate addresses

The big advantage of this is that we are able to update addresses over time and will always be able to reference our original addresses. Things like `Identify` will just be able to call this method when they need the latest addresses.

Reply:
Also with this, I think we can simplify this class; get rid of the getters and just expose the properties.

Reply:
This solution seems good to me and I will try it out. My main concern here is the same issue that I still have to solve: `noAnnounce` addresses using random ports for filtering. Unless the match criteria does not consider the port, we would need to make a comparison between the provided `listenAddr` and what is in `libp2p.transportManager.getAddrs()`

Reply:
changed accordingly

Reply:
> This solution seems good to me and I will try it out. My main concern here is the same issue that I still have to solve: noAnnounce addresses using random ports for filtering

Not sure I follow. Are you saying concerned that a listening address of `/ip4/0.0.0.0/tcp/0` can't be filtered with a noAnnounce of `/ip4/0.0.0.0/tcp/0`? If so, I don't think this matters and users shouldn't be trying to do this. While this is helpful for testing, listening on random ports is not recommended for production usage. The only noAnnounce "wildcard" I think would be nice to have is filtering loopback IP addresses, but that _could_ be checked by the app before configuring libp2p.

Reply:
Yes. I already added support for that now, but it is a bit messy 😅 

> The only noAnnounce "wildcard" I think would be nice to have is filtering loopback IP addresses, but that could be checked by the app before configuring libp2p.

👍